// @shardlabs/starknet-hardhat-plugin/dist/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkArtifactExists = exports.traverseFiles = exports.getDefaultHttpNetworkConfig = exports.adaptUrl = exports.adaptLog = void 0;
const plugins_1 = require("hardhat/plugins");
const constants_1 = require("./constants");
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const glob_1 = require("glob");
const util_1 = require("util");
const globPromise = (0, util_1.promisify)(glob_1.glob);
/**
 * Replaces Starknet specific terminology with the terminology used in this plugin.
 *
 * @param msg the log message to be adapted
 * @returns the log message with adaptation replacements
 */
function adaptLog(msg) {
    return msg
        .replace("--network", "--starknet-network")
        .replace("gateway_url", "gateway-url");
}
exports.adaptLog = adaptLog;
const DOCKER_HOST = "host.docker.internal";
/**
 * Adapts `url` by replacing localhost and 127.0.0.1 with `host.internal.docker`
 * @param url string representing the url to be adapted
 * @returns adapted url
 */
function adaptUrl(url) {
    if (process.platform === "darwin") {
        for (const protocol of ["http://", "https://", ""]) {
            for (const host of ["localhost", "127.0.0.1"]) {
                if (url === `${protocol}${host}`) {
                    return `${protocol}${DOCKER_HOST}`;
                }
                const prefix = `${protocol}${host}:`;
                if (url.startsWith(prefix)) {
                    return url.replace(prefix, `${protocol}${DOCKER_HOST}:`);
                }
            }
        }
    }
    return url;
}
exports.adaptUrl = adaptUrl;
function getDefaultHttpNetworkConfig(url) {
    return {
        url,
        accounts: undefined,
        gas: undefined,
        gasMultiplier: undefined,
        gasPrice: undefined,
        httpHeaders: undefined,
        timeout: undefined,
    };
}
exports.getDefaultHttpNetworkConfig = getDefaultHttpNetworkConfig;
function traverseFiles(traversable, fileCriteria = "*") {
    return __awaiter(this, void 0, void 0, function* () {
        let paths = [];
        if (fs.lstatSync(traversable).isDirectory()) {
            paths = yield globPromise(path.join(traversable, "**", fileCriteria));
        }
        else {
            paths.push(traversable);
        }
        const files = paths.filter(file => { return fs.lstatSync(file).isFile(); });
        return files;
    });
}
exports.traverseFiles = traverseFiles;
function checkArtifactExists(artifactsPath) {
    if (!fs.existsSync(artifactsPath)) {
        const msg = `Artifact expected to be at ${artifactsPath}, but not found. Consider recompiling your contracts.`;
        throw new plugins_1.HardhatPluginError(constants_1.PLUGIN_NAME, msg);
    }
}
exports.checkArtifactExists = checkArtifactExists;
